Bottom: 9c8c886a489050f572182efc391e495116753f62
Top:    a13237f65bb563f8ccc38e9bb3064518d9592ca5
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-08-13 11:06:49 -0500

Refresh of codebase-generator

---
diff --git a/docker/buildbot-master/buildbot/master.cfg b/docker/buildbot-master/buildbot/master.cfg
index 02056a49..e64f7121 100644
--- a/docker/buildbot-master/buildbot/master.cfg
+++ b/docker/buildbot-master/buildbot/master.cfg
@@ -139,3 +139,29 @@ c['workers'].extend(ptxdist.workers)
 c['schedulers'].extend(ptxdist.schedulers)
 c['builders'].extend(ptxdist.builders)
 c['change_source'].extend(ptxdist.change_source)
+
+from pprint import pprint
+from urlparse import urlparse
+def codebaseGenerator(chdict):
+    """ Generate a codebase description.
+    
+    The codebase generator takes information about a change, like repository,
+    branch, and revision, and creates a predictable string describing the code.
+    This is used to decide upon the builder for a change and also to tie a fork
+    repository back to its parent for merge requests.
+    """
+    pprint(chdict, indent=2)
+    path = urlparse(chdict['revlink']).path.lstrip('/')
+
+    components = []
+    while path != "":
+        path, tail = os.path.split(path)
+        if tail != "":
+            components.append(tail)
+    components.reverse()
+
+    codebase = os.path.join(*components[0:2])
+    return codebase
+
+
+c['codebaseGenerator'] = codebaseGenerator
